libr/anal/p/anal_avr.c

```diff
@@ -17,35 +17,277 @@
+#define INST_DECL(OP, M, SL, C, SZ, T)	{ #OP, (M), (SL), _inst__ ## OP, (C), (SZ), R_ANAL_OP_TYPE_ ## T }
+#define INST_LAST			{ "unknown", 0, 0, (void *) 0, 2, 1, R_ANAL_OP_TYPE_UNK      }
+
+#define INST_CALL(OPCODE_NAME)		_inst__ ## OPCODE_NAME (anal, op, addr, buf, len, fail, cpu)
+#define INST_INVALID			{ *fail = 1; return; }
+#define INST_ASSERT(x)			{ if (!(x)) { INST_INVALID; } }
+
+CPU_MODEL cpu_models[] = {
+	CPU_MODEL_DECL ("ATmega48",   11, 512),
+	CPU_MODEL_DECL ("ATmega8",    12, 512),
+	CPU_MODEL_DECL ("ATmega88",   12, 512),
+	CPU_MODEL_DECL ("ATmega168",  13, 512),
+	CPU_MODEL_DECL ("ATmega640",  16, 512),
+	CPU_MODEL_DECL ("ATmega1280", 16, 512),
+	CPU_MODEL_DECL ("ATmega1281", 16, 512),
+	CPU_MODEL_DECL ("ATmega2560", 22, 512),
+	CPU_MODEL_DECL ("ATmega2561", 22, 512),
+	CPU_MODEL_DECL ((char *) 0,   16, 512) };
+
+INST_HANDLER (cp) {
+	// CP Rd, Rr
+	int r = (buf[0]        & 0x0f) | ((buf[1] << 3) & 0x10);
+	int d = ((buf[0] >> 4) & 0x0f) | ((buf[1] << 4) & 0x10);
+
+	r_strbuf_setf (
+		&op->esil,
+		"r%d,r%d,==,"			// compare Rr vs Rd
+		"$z,zf,=,"			// zero flag (zf)
+		"r%d,r%d,^,0x08,!,!,hf,"	// half carry flag (hf)
+		"r%d,r%d,^,0x80,!,!,cf,"	// carry flag (cf)
+		"$o,vf,=,"			// overflow flag (vf)
+		"r%d,r%d,-,0x80,&,!,!,nf,=,"	// neg flg: result's MSB is set
+		"vf,nf,^,sf,=",			// sign flag: xor(vf, nf)
+		r, d, r, d, r, d, r, d);
+		//"r%d,r%d,==,"			// compare Rr vs Rd
+		//"$z,zf,=,"			// zero flag (zf)
+		//"$b3,hf,=,"			// half carry flag (hf)
+		//"$b8,cf,=,"			// carry flag (cf)
+		//"$o,vf,=,"			// overflow flag (vf)
+		//"r%d,r%d,-,0x80,&,!,!,nf,=,"	// neg flg: result's MSB is set
+		//"vf,nf,^,sf,=",			// sign flag: xor(vf, nf)
+}
+
+INST_HANDLER (cpc) {
+	int r = (buf[0]        & 0x0f) | ((buf[1] << 3) & 0x10);
+	int d = ((buf[0] >> 4) & 0x0f) | ((buf[1] << 4) & 0x10);
+
+	r_strbuf_setf (
+		&op->esil,
+		"r%d,cf,r%d,-,0xff,&,-,"	// ((Rd - C) & 0xff) - Rr
+		"0x80,&,!,!,nf,=,"		// nef flg: result's MSB is set
+		"r%d,cf,r%d,-,0xff,&,==,"	// ((Rd - C) & 0xff) == Rr
+		"$z,zf,=,"			// zero flag
+		"$b8,cf,=,"			// carry flag
+		"$b3,hf,=,"			// half carry flag
+		"$o,vf,=,"			// overflow flag
+		"vf,nf,^,sf,=",			// sign flag : xor(vf, nf)
+		r, d, r, d);
+}
+
+// ALIAS: Opcode (clr) when Rra and Rrb are the same
+INST_HANDLER (eor) {
+	int d = ((buf[0] & 0xf0) >> 4) | ((buf[1] & 1) << 4);
+	int r = (buf[0] & 0xf) | ((buf[1] & 2) << 3);
+
+	r_strbuf_setf (&op->esil, "r%d,r%d,^=,$z,zf,=,r%d,0x80,&,!,!,nf,=,nf,sf,=,0,vf,=", r, d, d);
+}
+
+INST_HANDLER (movw) {
+	int d = (buf[0] & 0xf0) >> 3;
+	int r = (buf[0] & 0x0f) << 1;
+
+	r_strbuf_setf (&op->esil, "r%d,r%d,=,r%d,r%d,=", r, d, r + 1, d + 1);
+}
+
+INST_HANDLER (nop) {
+}
+
+INST_HANDLER (out) {
+	op->type2 = 1;
+	op->val = (buf[0] & 0x0f) | (((buf[1] >> 1) & 0x03) << 4);
+	
+	int r = ((buf[0] >> 4) & 0x0f) | ((buf[1] & 0x01) << 4);
```

Comment:
can we have this in the first line? it is good practice to define variables at the begining of the function